lib/concurrent/configuration.rb

```diff
@@ -1,16 +1,21 @@
 require 'thread'
-require 'concurrent/atomics'
 require 'concurrent/errors'
-require 'concurrent/executors'
 require 'concurrent/concern/deprecation'
 require 'concurrent/concern/logging'
-require 'concurrent/utility/at_exit'
-require 'concurrent/utility/processor_counter'
 
 module Concurrent
   extend Concern::Logging
   extend Concern::Deprecation
 
+  autoload :AtExit,             'concurrent/utility/at_exit'
+  autoload :AtomicReference,    'concurrent/atomic/atomic_reference'
+  autoload :Delay,              'concurrent/delay'
+  autoload :ProcessorCounter,   'concurrent/utility/processor_counter'
+  autoload :TimerSet,           'concurrent/executor/timer_set'
+  autoload :ImmediateExecutor,  'concurrent/executor/immediate_executor'
```

Comment:
Here's the response: "autoload itself is safe but class definition happening in parallel can still expose incomplete classes."

https://twitter.com/headius/status/626797364066578432